<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API Reference · SMGReader.jl</title><meta name="title" content="API Reference · SMGReader.jl"/><meta property="og:title" content="API Reference · SMGReader.jl"/><meta property="twitter:title" content="API Reference · SMGReader.jl"/><meta name="description" content="Documentation for SMGReader.jl."/><meta property="og:description" content="Documentation for SMGReader.jl."/><meta property="twitter:description" content="Documentation for SMGReader.jl."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SMGReader.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">SMGReader.jl</a></li><li><a class="tocitem" href="../chromstencil/">Chromatin Stencilling</a></li><li><a class="tocitem" href="../directrna/">Direct RNA</a></li><li class="is-active"><a class="tocitem" href>API Reference</a><ul class="internal"><li><a class="tocitem" href="#External-API"><span>External API</span></a></li><li><a class="tocitem" href="#Internal-API"><span>Internal API</span></a></li><li><a class="tocitem" href="#Base-Extensions"><span>Base Extensions</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API Reference</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/exetergrdd/SMGReader.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/exetergrdd/SMGReader.jl/blob/main/docs/src/api.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h1><ul><li><a href="#External-API">External API</a></li><li><a href="#Internal-API">Internal API</a></li><li><a href="#Base-Extensions">Base extensions</a></li></ul><h2 id="External-API"><a class="docs-heading-anchor" href="#External-API">External API</a><a id="External-API-1"></a><a class="docs-heading-anchor-permalink" href="#External-API" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.AuxField" href="#SMGReader.AuxField"><code>SMGReader.AuxField</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">AuxField</code></pre><p>Describes the location and type of an auxiliary field in a BAM record.</p><p>Fields:</p><ul><li><code>tag::NTuple{2, UInt8}</code>: Two-character tag identifying the field.</li><li><code>typechar::UInt8</code>: Type character for the field (e.g., &#39;A&#39;, &#39;Z&#39;, &#39;B&#39;, etc.).</li><li><code>elemtypechar::UInt8</code>: Element type character for array fields (&#39;B&#39;).</li><li><code>start::Int</code>: Start index (1-based) of the field value in the record&#39;s byte array.</li><li><code>stop::Int</code>: Stop index (1-based) of the field value in the record&#39;s byte array.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/auxfields.jl#L12-L23">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.AuxFieldIter" href="#SMGReader.AuxFieldIter"><code>SMGReader.AuxFieldIter</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">AuxFieldIter(record::BamRecord)</code></pre><p>An iterator over the auxiliary fields of a BAM record.  Yields <code>AuxField</code> objects describing the location and type of each aux field in the record.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/auxfields.jl#L1-L6">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.AuxMapMod" href="#SMGReader.AuxMapMod"><code>SMGReader.AuxMapMod</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">AuxMapMod()</code></pre><p>Construct an map of the aux data for HP, MM, ML fields. For standard use in chromatin stencilling (non-fire) and direct RNA.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/auxfields.jl#L134-L138">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.AuxMapModFire" href="#SMGReader.AuxMapModFire"><code>SMGReader.AuxMapModFire</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">AuxMapModFire()</code></pre><p>Construct and map of the aux data for a FIRE BAM/CRAM mapping fields</p><ul><li>hp: Haplotype</li><li>mm: Run length encoded mod bases</li><li>ml: Mod Probability</li><li>ns: Nucleosome positions</li><li>nl: nucleosome length</li><li>as: msp positions</li><li>al: msp lengths</li><li>aq: msp quality score</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/auxfields.jl#L147-L160">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.DirectRNA" href="#SMGReader.DirectRNA"><code>SMGReader.DirectRNA</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">DirectRNA(auxmap::AuxMap)</code></pre><p>Construct preallocated HTS read data building a read to genome map for DirectRNA setting an BAM auxillary data map <code>AuxMap</code>. e.g.</p><ul><li>For generic tracking of modifications <code>DirectRNA(AuxMapMod())</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L26-L33">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.DirectRNAAlignBlocks" href="#SMGReader.DirectRNAAlignBlocks"><code>SMGReader.DirectRNAAlignBlocks</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">DirectRNAAlignBlocks(auxmap::AuxMap)</code></pre><p>Construct preallocated HTS read data building both a read to genome map and an align blocks to genome map for spliced alignments.  e.g.</p><ul><li>For generic tracking of modifications <code>DirectRNAAlignBlocks(AuxMapMod())</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L41-L48">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.ModIterator" href="#SMGReader.ModIterator"><code>SMGReader.ModIterator</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">struct ModIterator</code></pre><p>An iterator over all modification calls (as <code>ModificationInfo</code>) in a BAM/CRAM record, as described by the MM and ML auxiliary fields.</p><p><strong>Fields</strong></p><ul><li><code>record::BamRecord</code>: The BAM/CRAM record.</li><li><code>mmaux::AuxField</code>: The MM auxiliary field (modification code and run length encoding of).</li><li><code>mlaux::AuxField</code>: The ML auxiliary field (modification probabilities).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L248-L257">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.ModIterator-Tuple{BamRecord, SMGReader.HTSReadData}" href="#SMGReader.ModIterator-Tuple{BamRecord, SMGReader.HTSReadData}"><code>SMGReader.ModIterator</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ModIterator(record::BamRecord, rdata::HTSReadData)</code></pre><p>Construct a <code>ModIterator</code> for a given BAM/CRAM record, using the mm and ml fields in the auxmap</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L274-L278">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.ModIterator-Tuple{BamRecord}" href="#SMGReader.ModIterator-Tuple{BamRecord}"><code>SMGReader.ModIterator</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ModIterator(record::BamRecord)</code></pre><p>Construct a <code>ModIterator</code> for a given BAM/CRAM record, automatically extracting the MM and ML fields.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L264-L268">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.Modification" href="#SMGReader.Modification"><code>SMGReader.Modification</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Modification</code></pre><p>Enumerates supported DNA/RNA modification types, such as 6mA, 5mC, 5hmC, inosine, pseudouridine, and others. Used to identify the type of modification parsed from MM/ML fields.</p><p>Currently supported mod codes:</p><table><tr><th style="text-align: right">Enum value</th><th style="text-align: right">Modification</th><th style="text-align: right">BAM code</th></tr><tr><td style="text-align: right"><code>mod_6mA</code></td><td style="text-align: right">N6-methyladenine</td><td style="text-align: right"><code>a</code></td></tr><tr><td style="text-align: right"><code>mod_5mC</code></td><td style="text-align: right">5-methylcytosine</td><td style="text-align: right"><code>m</code></td></tr><tr><td style="text-align: right"><code>mod_5hmC</code></td><td style="text-align: right">5-hydroxymethylcytosine</td><td style="text-align: right"><code>h</code></td></tr><tr><td style="text-align: right"><code>mod_inosine</code></td><td style="text-align: right">Inosine</td><td style="text-align: right"><code>17596</code></td></tr><tr><td style="text-align: right"><code>mod_pseU</code></td><td style="text-align: right">Pseudouridine</td><td style="text-align: right"><code>17802</code></td></tr><tr><td style="text-align: right"><code>mod_4mC</code></td><td style="text-align: right">N4-methylcytosine</td><td style="text-align: right"><code>21839</code></td></tr></table></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L3-L19">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.ModificationInfo" href="#SMGReader.ModificationInfo"><code>SMGReader.ModificationInfo</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ModificationInfo</code></pre><p>Holds information about a single modification call parsed from a BAM/CRAM record.</p><p><strong>Fields</strong></p><ul><li><code>mod::Modification</code>: The modification type.</li><li><code>strand::Bool</code>: <code>true</code> for forward (+), <code>false</code> for reverse (-) strand.</li><li><code>pos::Int</code>: 1-based position in the read sequence. </li><li><code>prob::UInt8</code>: Modification probability (from ML field).</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L230-L240">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.StencillingData" href="#SMGReader.StencillingData"><code>SMGReader.StencillingData</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">StencillingData(auxmap::AuxMap)</code></pre><p>Construct preallocated HTS read data building a read to genome map for chromatin stencilling setting an BAM auxillary data map <code>AuxMap</code>. e.g.</p><ul><li>For generic tracking of modifications <code>StencillingData(AuxMapMod())</code></li><li>For tracking FIRE fields <code>StencillingData(AuxMapModFire())</code></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L11-L19">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.VectorBuffer" href="#SMGReader.VectorBuffer"><code>SMGReader.VectorBuffer</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mutable struct VectorBuffer{T} &lt;: AbstractVector{T}</code></pre><p>A resizable buffer that wraps a <code>Vector{T}</code> and tracks a logical length (<code>len</code>). Efficient for repeated resizing and reuse, e.g., when processing records in a loop.</p><pre><code class="nohighlight hljs">Recommendation is to allocate a large enough vector buffer to work with, e.g.
following constants are set in processedread.jl

const MOD_6MA_BUFFER_LENGTH = 70_000
const MOD_5MC_BUFFER_LENGTH = 70_000
const READ_BUFFER_LENGTH = 200_000
const ALIGN_BLOCK_LENGTH = 5_000</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/buffer.jl#L1-L14">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.VectorBuffer-Union{Tuple{Int64}, Tuple{T}, Tuple{Int64, Any}} where T" href="#SMGReader.VectorBuffer-Union{Tuple{Int64}, Tuple{T}, Tuple{Int64, Any}} where T"><code>SMGReader.VectorBuffer</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">VectorBuffer{T}(backcapacity::Int, initcapacity=0)</code></pre><p>Construct a <code>VectorBuffer</code> with backing storage of size <code>backcapacity</code> and logical length <code>initcapacity</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/buffer.jl#L20-L24">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.cigarstring-Tuple{Any}" href="#SMGReader.cigarstring-Tuple{Any}"><code>SMGReader.cigarstring</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">cigarstring(record)</code></pre><p>Return CIGAR string of <code>record</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L60-L64">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.cigarvec" href="#SMGReader.cigarvec"><code>SMGReader.cigarvec</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">cigarvec(record)</code></pre><p>Get CIGAR description of alignment of <code>record</code> returned as lengths (Vector{Int}) and codes (Vector{Char}).</p><p>To build alignment map use <code>readtogenome</code> instead of working with cigar directly</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L34-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.eachintersection-Tuple{HTSFileReader, AbstractString, UnitRange{Int64}}" href="#SMGReader.eachintersection-Tuple{HTSFileReader, AbstractString, UnitRange{Int64}}"><code>SMGReader.eachintersection</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">eachintersection(reader::HTSFileReader, chrom::AbstractString, loc::UnitRange{Int})</code></pre><p>Construct an iterator of <code>reader</code> over <code>chrom</code> and <code>loc</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htslib.jl#L251-L255">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.eachrecord-Tuple{HTSFileReader}" href="#SMGReader.eachrecord-Tuple{HTSFileReader}"><code>SMGReader.eachrecord</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">eachrecord(reader::HTSFileReader)</code></pre><p>Construct an iterator of <code>reader</code> in record order. Multiple successive calls will continue iterating from current point in reader, resetting iterator requires closing and opening file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htslib.jl#L212-L216">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firegenome-Tuple{BamRecord, StencillingData{AuxMapModFire}}" href="#SMGReader.firegenome-Tuple{BamRecord, StencillingData{AuxMapModFire}}"><code>SMGReader.firegenome</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firegenome(record::BamRecord, recorddata::StencillingData{AuxMapModFire}</code></pre><p>Construct a function for mapping fire nucleosomes and MSP coordiates from read coordinates to genome filtering out entries that do not map</p><p>Usage:     firegenomecoords = firegenome(record, recorddata)</p><pre><code class="nohighlight hljs">for (pos, len, qual) in firegenomcoords(firemsps(record, recorddata))
    ### do something with genome coordinates
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L279-L290">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firegenomecoords-NTuple{4, Any}" href="#SMGReader.firegenomecoords-NTuple{4, Any}"><code>SMGReader.firegenomecoords</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firegenomecoords(pos::Int, len::Int, record::BamRecord, recorddata::HTSReadData; onebased=true)</code></pre><p>Special version of <code>genomecoords</code> for fiber-tools and fire elements that are encoded 0-based.</p><ul><li><code>pos</code> is zero-based position on the read</li><li><code>one-based = true</code>  returns one based inclusive coords</li><li><code>one-based = false</code> returns zero based exclusive coords</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L192-L201">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firemsps-Tuple{BamRecord, StencillingData{AuxMapModFire}}" href="#SMGReader.firemsps-Tuple{BamRecord, StencillingData{AuxMapModFire}}"><code>SMGReader.firemsps</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firemsps(r::BamRecord, recorddata::StencillingData{AuxMapModFire})</code></pre><p>Construct iterator of fire msp positions, lengths, and quality scores in read coordinates Usage:</p><pre><code class="nohighlight hljs">for (pos, len, qual) in firemsps(record, recorddata)
    ### do something 
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L249-L258">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firenucs-Tuple{BamRecord, StencillingData{AuxMapModFire}}" href="#SMGReader.firenucs-Tuple{BamRecord, StencillingData{AuxMapModFire}}"><code>SMGReader.firenucs</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firenucs(r::BamRecord, recorddata::StencillingData{AuxMapModFire})</code></pre><p>Construct iterator of fire nucleosome positions, lengths in read coordinates Usage:</p><pre><code class="nohighlight hljs">for (pos, len) in firenucs(record, recorddata)
    ### do something 
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L261-L270">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.flag-Tuple{BamRecord}" href="#SMGReader.flag-Tuple{BamRecord}"><code>SMGReader.flag</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">flag(record::BamRecord)</code></pre><p>Get flag of record</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L131-L135">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.genomecoords-Tuple{Any, Any, BamRecord, SMGReader.HTSReadData}" href="#SMGReader.genomecoords-Tuple{Any, Any, BamRecord, SMGReader.HTSReadData}"><code>SMGReader.genomecoords</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">genomecoords(pos::Int, len::Int, record::BamRecord, recorddata::HTSReadData; onebased=true)</code></pre><p>Map read coordinates to genome coordates that uses <code>alignmap</code> in <code>recorddata</code>,</p><ul><li><code>pos</code> is one-based to match the alignment map</li><li><code>one-based = true</code>  returns one based inclusive coords</li><li><code>one-based = false</code> returns zero based exclusive coords</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L116-L125">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.genomecoords-Tuple{Int64, BamRecord, SMGReader.HTSReadData}" href="#SMGReader.genomecoords-Tuple{Int64, BamRecord, SMGReader.HTSReadData}"><code>SMGReader.genomecoords</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">genomecoords(pos::Int, record::BamRecord, recorddata::HTSReadData; onebased=true)</code></pre><p>Map read coordinates to genome coordates that uses <code>alignmap</code> in <code>recorddata</code>, <code>onebased=true</code> for 1-based coordinates</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L102-L106">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.haplotype" href="#SMGReader.haplotype"><code>SMGReader.haplotype</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">haplotype(record::BamRecord, recorddata::HTSReadData)</code></pre><p>Return haplotype of record and nothing if haplotype missing</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L85-L89">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.hashaplotype-Tuple{BamRecord, SMGReader.HTSReadData}" href="#SMGReader.hashaplotype-Tuple{BamRecord, SMGReader.HTSReadData}"><code>SMGReader.hashaplotype</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">hashaplotype(record::BamRecord, recorddata::HTSReadData)</code></pre><p>Return true if <code>record</code> has haplotype field</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L92-L96">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.indexfile-Tuple{Any}" href="#SMGReader.indexfile-Tuple{Any}"><code>SMGReader.indexfile</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">indexfile(file)</code></pre><p>Return a <code>.bai</code> or <code>.crai</code> file as appropriate for <code>bam</code> and <code>cram</code> files</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htslib.jl#L99-L103">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.ispositive-Tuple{BamRecord}" href="#SMGReader.ispositive-Tuple{BamRecord}"><code>SMGReader.ispositive</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ispositive(record::BamRecord)</code></pre><p>Determine if <code>record</code> is aligned to the positive strand</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L138-L142">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.leftposition-Tuple{BamRecord}" href="#SMGReader.leftposition-Tuple{BamRecord}"><code>SMGReader.leftposition</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">leftposition(record::BamRecord)</code></pre><p>Return the position of the record in 0-based coords</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L154-L158">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.mappingquality-Tuple{BamRecord}" href="#SMGReader.mappingquality-Tuple{BamRecord}"><code>SMGReader.mappingquality</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mappingquality(record::BamRecord)</code></pre><p>Return mapping quality of record</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L162-L166">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.matepos-Tuple{BamRecord}" href="#SMGReader.matepos-Tuple{BamRecord}"><code>SMGReader.matepos</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">mateposition(record::BamRecord)</code></pre><p>return the mate position in 0-based coords</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L169-L172">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.materefname-Tuple{HTSFileReader, BamRecord}" href="#SMGReader.materefname-Tuple{HTSFileReader, BamRecord}"><code>SMGReader.materefname</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">materefname(reader::HTSFileReader, record::BamRecord)</code></pre><p>Return the reference name of the mate of <code>record</code> read by <code>reader</code>. <code>reader</code> is necessary to load the ref name from the header of the bam/cram file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L27-L31">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.processread!-Tuple{BamRecord, SMGReader.HTSReadData}" href="#SMGReader.processread!-Tuple{BamRecord, SMGReader.HTSReadData}"><code>SMGReader.processread!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">processread!(record::BamRecord, recorddata::HTSReadData)</code></pre><p>Process read mapping auxillary fields and building read to genome map using preallocated <code>recorddata</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L58-L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.qname-Tuple{BamRecord}" href="#SMGReader.qname-Tuple{BamRecord}"><code>SMGReader.qname</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">qname(record::BamRecord)</code></pre><p>Return query name of <code>record</code>. The read name.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L4-L8">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.qual-Tuple{BamRecord}" href="#SMGReader.qual-Tuple{BamRecord}"><code>SMGReader.qual</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">qual(record::BamRecord)</code></pre><p>Get quality scores as byte array</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L110-L114">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.querylength-Tuple{BamRecord}" href="#SMGReader.querylength-Tuple{BamRecord}"><code>SMGReader.querylength</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">querylength(record::BamRecord)</code></pre><p>Return the query length of the record</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L183-L187">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.refname-Tuple{HTSFileReader, BamRecord}" href="#SMGReader.refname-Tuple{HTSFileReader, BamRecord}"><code>SMGReader.refname</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">refname(reader::HTSFileReader, record::BamRecord)</code></pre><p>Return the reference name of <code>record</code> read by <code>reader</code>. <code>reader</code> is necessary to load the ref name from the header of the bam/cram file.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L13-L17">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.rightposition-Tuple{SMGReader.HTSReadData}" href="#SMGReader.rightposition-Tuple{SMGReader.HTSReadData}"><code>SMGReader.rightposition</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">rightposition(recorddata::HTSReadData)</code></pre><p>Get the right position of alignment on genome.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L77-L81">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.seq-Tuple{Any}" href="#SMGReader.seq-Tuple{Any}"><code>SMGReader.seq</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">seq(record)</code></pre><p>Return seq string of <code>record</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L77-L81">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.setlength!" href="#SMGReader.setlength!"><code>SMGReader.setlength!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">setlength!(buff::VectorBuffer, newlen, backfactor=1.2)</code></pre><p>Resize the logical length of the buffer to <code>newlen</code>. If the backing storage is too small, resize it by a factor of <code>backfactor</code> or double the current length, whichever is larger. Returns the buffer.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/buffer.jl#L35-L41">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.templatelength-Tuple{BamRecord}" href="#SMGReader.templatelength-Tuple{BamRecord}"><code>SMGReader.templatelength</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">templatelength(record::BamRecord)</code></pre><p>return the template length of the record</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L176-L180">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.validflag-Tuple{BamRecord}" href="#SMGReader.validflag-Tuple{BamRecord}"><code>SMGReader.validflag</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">validfrag(record::BamRecord)</code></pre><p>Return true if flag is 0 or 16.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htsrecord.jl#L145-L149">source</a></section></article><h2 id="Internal-API"><a class="docs-heading-anchor" href="#Internal-API">Internal API</a><a id="Internal-API-1"></a><a class="docs-heading-anchor-permalink" href="#Internal-API" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.ModIteratorState" href="#SMGReader.ModIteratorState"><code>SMGReader.ModIteratorState</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ModIteratorState

Mutable internal state of ModIterator</code></pre><p><strong>Fields</strong></p><ul><li><code>newrecord::Bool</code>: Bool marking whether next iteration will return a new modification with different modcode</li><li><code>base::UInt8</code>: Current base associated with modcode</li><li><code>strand::Bool</code>: String of modfication</li><li><code>mod::Modification</code>: Current modification</li><li><code>mm_pos::Int</code>: Position in the MM string</li><li><code>ml_pos::Int</code>: Position in the ML vector</li><li><code>seq_pos::Int</code>: Position in the bam sequence</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L282-L295">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.alignmentblocks!-Tuple{BamRecord, VectorBuffer{SMGReader.AlignBlock}}" href="#SMGReader.alignmentblocks!-Tuple{BamRecord, VectorBuffer{SMGReader.AlignBlock}}"><code>SMGReader.alignmentblocks!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">alignmentblocks!(record::BamRecord, alignblocks::VectorBuffer{AlignBlock})</code></pre><p>In place update of alignment blocks for a <code>record</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/alignmentmap.jl#L78-L82">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.auxmap!-Tuple{BamRecord, AuxMapMod}" href="#SMGReader.auxmap!-Tuple{BamRecord, AuxMapMod}"><code>SMGReader.auxmap!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">auxmap!(record::BamRecord, auxmap&lt;:Auxmap)</code></pre><p>Construct map of auxillary data from <code>record</code> identifying fields specified by <code>auxmap</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/auxfields.jl#L174-L178">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.checkbounds-Tuple{VectorBuffer, Any}" href="#SMGReader.checkbounds-Tuple{VectorBuffer, Any}"><code>SMGReader.checkbounds</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkbounds(buff::VectorBuffer, i)</code></pre><p>Check if index <code>i</code> is within the logical bounds of the buffer. Throws <code>BoundsError</code> if out of bounds.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/buffer.jl#L51-L56">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.findnextbase-Tuple{BamRecord, UInt8, Int64, Int64}" href="#SMGReader.findnextbase-Tuple{BamRecord, UInt8, Int64, Int64}"><code>SMGReader.findnextbase</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">findnextbase(record::BamRecord, base::UInt8, start::Int, skip::Int)</code></pre><p>Find the next occurence of <code>base</code> = (0x01, 0x02, 0x04, 0x08)<code>for</code>(A, C, G, T)<code>within the bam bit packed sequence, starting from position</code>start<code>skipping over</code>skip` occurences of the base.</p><p>This function is designed to locate the next base for which modification data is available for parsing the MM run length encoded aux field.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L115-L122">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.findprevbase-Tuple{BamRecord, UInt8, Int64, Int64}" href="#SMGReader.findprevbase-Tuple{BamRecord, UInt8, Int64, Int64}"><code>SMGReader.findprevbase</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">findprevbase(record::BamRecord, base::UInt8, start::Int, skip::Int)</code></pre><p>Find the prev occurence of <code>base</code> = (0x01, 0x02, 0x04, 0x08)<code>for</code>(A, C, G, T)<code>within the bam bit packed sequence, starting from position</code>start<code>skipping over</code>skip` occurences of the base.</p><p>This function is designed to locate the prev base for which modification data is available for parsing the MM run length encoded aux field.</p><p>This is the reverse version of <code>findnextbase</code> for use in reverse strand aligned reads.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L170-L179">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firemsplen-Tuple{BamRecord, StencillingData{AuxMapModFire}}" href="#SMGReader.firemsplen-Tuple{BamRecord, StencillingData{AuxMapModFire}}"><code>SMGReader.firemsplen</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firemsplen(r::BamRecord, recorddata::StencillingData{AuxMapModFire})</code></pre><p>Return fiber tools annotated MSP lengths on read coordinates.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L227-L231">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firemsppos-Tuple{BamRecord, StencillingData{AuxMapModFire}}" href="#SMGReader.firemsppos-Tuple{BamRecord, StencillingData{AuxMapModFire}}"><code>SMGReader.firemsppos</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firemsppos(r::BamRecord, recorddata::StencillingData{AuxMapModFire})</code></pre><p>Return fiber tools annotated MSP positions on read coordinates. 0-based coordinates.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L220-L224">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firemspqual-Tuple{BamRecord, StencillingData{AuxMapModFire}}" href="#SMGReader.firemspqual-Tuple{BamRecord, StencillingData{AuxMapModFire}}"><code>SMGReader.firemspqual</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firemspqual(r::BamRecord, recorddata::StencillingData{AuxMapModFire})</code></pre><p>Return fiber tools annotated MSP quality scores on read coordinates.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L234-L238">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firenuclen-Tuple{BamRecord, StencillingData{AuxMapModFire}}" href="#SMGReader.firenuclen-Tuple{BamRecord, StencillingData{AuxMapModFire}}"><code>SMGReader.firenuclen</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firenuclen(r::BamRecord, recorddata::StencillingData{AuxMapModFire})</code></pre><p>Return fiber tools annotated nucleosome lengths on read coordinates.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L213-L217">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.firenucpos-Tuple{BamRecord, StencillingData{AuxMapModFire}}" href="#SMGReader.firenucpos-Tuple{BamRecord, StencillingData{AuxMapModFire}}"><code>SMGReader.firenucpos</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">firenucpos(r::BamRecord, recorddata::StencillingData{AuxMapModFire})</code></pre><p>Return fiber tools annotated nucleosome positions on read coordinates 0-based coordinates.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/processedread.jl#L206-L210">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.getmmml-Tuple{BamRecord}" href="#SMGReader.getmmml-Tuple{BamRecord}"><code>SMGReader.getmmml</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getmmml(record::BamRecord)</code></pre><p>Function to retrieve location of MM and ML auxillary fields as <code>AuxFields</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/auxfields.jl#L105-L109">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.getmodbasestrand-Tuple{UInt8, Bool}" href="#SMGReader.getmodbasestrand-Tuple{UInt8, Bool}"><code>SMGReader.getmodbasestrand</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">getmodbasestrand(b::UInt8, ispos::Bool) -&gt; UInt8</code></pre><p>Given a base character and strand, return the encoded base for modification search.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L36-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.parse_mod_code-Tuple{Any, Any}" href="#SMGReader.parse_mod_code-Tuple{Any, Any}"><code>SMGReader.parse_mod_code</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">parse_mod_code(data, i)</code></pre><p>Parse the modcode at position <code>i</code> in bytearray <code>data</code> Return <code>(mod, next_i)</code> where <code>mod</code> is a <code>Modifcation</code> and <code>next_i</code> is the next index after the modcode in the bytearray.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L58-L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="SMGReader.readtogenome!-Union{Tuple{T}, Tuple{BamRecord, VectorBuffer{T}}} where T" href="#SMGReader.readtogenome!-Union{Tuple{T}, Tuple{BamRecord, VectorBuffer{T}}} where T"><code>SMGReader.readtogenome!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readtogenome!(record::BamRecord, alignmap::VectorBuffer{T}) where {T}</code></pre><p>In place update of <code>alignmap</code> VectorBuffer with basepair level read to genome map. Uses CIGAR information in <code>record</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/alignmentmap.jl#L2-L7">source</a></section></article><h2 id="Base-Extensions"><a class="docs-heading-anchor" href="#Base-Extensions">Base Extensions</a><a id="Base-Extensions-1"></a><a class="docs-heading-anchor-permalink" href="#Base-Extensions" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.close-Tuple{HTSFileReader}" href="#Base.close-Tuple{HTSFileReader}"><code>Base.close</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">close(reader::HTSFileReader)</code></pre><p>Close open <code>reader</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htslib.jl#L146-L150">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.getindex-Union{Tuple{T}, Tuple{VectorBuffer{T}, AbstractVector{Bool}}} where T" href="#Base.getindex-Union{Tuple{T}, Tuple{VectorBuffer{T}, AbstractVector{Bool}}} where T"><code>Base.getindex</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.getindex(buff::VectorBuffer{T}, inds::AbstractVector{Bool})</code></pre><p>Return a regular array of elements where <code>inds</code> is <code>true</code>. Logical indexing must match the logical length of the buffer.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/buffer.jl#L85-L90">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.getindex-Union{Tuple{T}, Tuple{VectorBuffer{T}, Int64}} where T" href="#Base.getindex-Union{Tuple{T}, Tuple{VectorBuffer{T}, Int64}} where T"><code>Base.getindex</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.getindex(buff::VectorBuffer{T}, i::Int)</code></pre><p>Return the element at logical index <code>i</code> in the buffer.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/buffer.jl#L65-L69">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.getindex-Union{Tuple{T}, Tuple{VectorBuffer{T}, UnitRange{Int64}}} where T" href="#Base.getindex-Union{Tuple{T}, Tuple{VectorBuffer{T}, UnitRange{Int64}}} where T"><code>Base.getindex</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.getindex(buff::VectorBuffer{T}, r::UnitRange{Int})</code></pre><p>Return a slice of the buffer for the given range <code>r</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/buffer.jl#L96-L100">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.iterate-Tuple{AuxFieldIter}" href="#Base.iterate-Tuple{AuxFieldIter}"><code>Base.iterate</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.iterate(iter::AuxFieldIter)
Base.iterate(iter::AuxFieldIter, i)</code></pre><p>Iterate over the auxiliary fields in a BAM record.  Returns a tuple <code>(AuxField, next_index)</code> for each field, or <code>nothing</code> when done.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/auxfields.jl#L34-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.iterate-Tuple{ModIterator}" href="#Base.iterate-Tuple{ModIterator}"><code>Base.iterate</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.iterate(iter::ModIterator)

Iterate through modifications in order they are present in BAM record returning `ModInfo` for each.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/modifications.jl#L309-L313">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.open-Tuple{Type{HTSFileReader}, Any}" href="#Base.open-Tuple{Type{HTSFileReader}, Any}"><code>Base.open</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">open(HTSFileReader, path; idx_path=indexfile(path), bamthreads=Threads.nthreads())</code></pre><p>Use ccall to <code>hts_open</code> the file at <code>path</code>.     - <code>idx_path</code> gives location of index and is mandatory     - <code>bamthreads</code> specifies number of threads used by htslib</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htslib.jl#L118-L124">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.read!-Tuple{HTSFileReader, BamRecord}" href="#Base.read!-Tuple{HTSFileReader, BamRecord}"><code>Base.read!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.read!(reader::HTSFileReader, record::BamRecord)</code></pre><p>Read next record in file order from <code>reader</code> and in place store result in <code>record</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htslib.jl#L167-L171">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.read-Tuple{HTSFileReader}" href="#Base.read-Tuple{HTSFileReader}"><code>Base.read</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.read(reader::HTSFileReader)</code></pre><p>Allocating version of <code>read!</code> Read next record in file order from <code>reader</code> return a <code>record</code>, recommended to use <code>read!</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/htslib.jl#L188-L193">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.setindex!-Union{Tuple{T}, Tuple{VectorBuffer{T}, Any, Int64}} where T" href="#Base.setindex!-Union{Tuple{T}, Tuple{VectorBuffer{T}, Any, Int64}} where T"><code>Base.setindex!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.setindex!(buff::VectorBuffer{T}, val, i::Int)</code></pre><p>Set the element at logical index <code>i</code> to <code>val</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/exetergrdd/SMGReader.jl/blob/31071e91010e6ca92a44a7a28db6278b47701f24/src/buffer.jl#L75-L79">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../directrna/">« Direct RNA</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Thursday 11 September 2025 14:43">Thursday 11 September 2025</span>. Using Julia version 1.11.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
